friendshipper/src-tauri/src/system/update.rs

```diff
@@ -1,107 +1,161 @@
-
-        let key = latest.key.to_string();
-
-        info!("Downloading key: {}", key);
-
-        let mut file = File::create(tmp_path.clone())?;
-        let config = state.app_config.read().clone();
-        let mut result = match client
-            .get_object()
-            .bucket(config.aws_config.unwrap().artifact_bucket_name)
-            .key(key)
-            .send()
-            .await
-        {
-            Ok(result) => result,
-            Err(e) => {
-                error!("Error downloading exe: {:?}", e);
-                return Err(anyhow!("Error downloading exe: {:?}", e).into());
-            }
-        };
-
-        while let Some(bytes) = result.body.try_next().await? {
-            file.write_all(&bytes)?;
-        }
-
-        return match self_replace::self_replace(&tmp_path) {
-            Ok(_) => {
-                info!("Removing file: {}", tmp_path);
-                fs::remove_file(&tmp_path)?;
 
-                info!("Update complete");
-                Ok(())
-            }
-            Err(e) => {
-                error!("Error replacing exe: {:?}", e);
-                Err(anyhow!("Error replacing exe: {:?}", e).into())
+            if release.tag_name.starts_with(APP_NAME)
+                && release
+                    .assets
+                    .iter()
+                    .any(|asset| asset.name == format!("{}{}", APP_NAME, BIN_SUFFIX))
+            {
+                return Some(release);
             }
-        };
-    }
-
-    Err(anyhow!("No latest version found").into())
-}
-
-pub fn get_data_dir() -> Option<PathBuf> {
-    if let Some(base_dirs) = BaseDirs::new() {
-        let data_dir = base_dirs.data_dir().join(APP_NAME).join("data");
-
-        fs::create_dir_all(&data_dir).unwrap();
 
-        Some(data_dir)
-    } else {
-        None
+            None
+        })
+        .next();
+
+    match latest {
+        Some(latest) => {
+            let asset = latest
+                .assets
+                .iter()
+                .find(|asset| asset.name == format!("{}{}", APP_NAME, BIN_SUFFIX));
+
+            match asset {
+                Some(asset) => {
+                    // download release
+                    info!(
+                        "https://api.github.com/repos/{}/{}/releases/assets/{}",
+                        REPO_OWNER, REPO_NAME, asset.id
+                    );
+
+                    let client = reqwest::Client::new();
+                    let mut response = client
+                        .get(format!(
+                            "https://api.github.com/repos/{}/{}/releases/assets/{}",
+                            REPO_OWNER, REPO_NAME, asset.id
+                        ))
+                        .header("Authorization", format!("Bearer {}", token))
+                        .header("Accept", "application/octet-stream")
+                        .header("User-Agent", APP_NAME)
+                        .send()
+                        .await?;
```

Comment:
I wonder if it's possible to get a callback while the stream is going to track overall progress so we could display a modal. Maybe later though.